allow automatic simplification of sqrt if result can be written without sqrt

smichr opened the pull request:
`sqrt(a + I*b)` can sometimes be written simply as `c + I*d`. This PR checks for those conditions. Tests are added that demonstrate the new capability. The principle value of the `sqrt` is returned (see Algebraic Formula [here](https://en.wikipedia.org/w/index.php?title=Square_root)).

- [x] have all modifications recognize (a+I*b)**(n/2) where n is an integer and return the expanded result, e.g. `(3+4*I)**(S(3)/2)` is `2+11*I`
- [x] handle `sqrt(-2*I)` in a fashion similar to `sqrt(2*I)` except result will be `r - I*r`

closes #12739

smichr commented:
@asmeurer , any objections or should this be part of `_simplify_basic` for power (or whatever the basic simplification method is)? Or maybe part of `re` and `im`? Right now, this sort of simplification is not recognized:

```
>>> sqrt(32 - 24*I)
sqrt(32 - 24*I)
>>> simplify(_)
2*sqrt(8 - 6*I)  # could be 6 - 2*I
>>> _.n()
6.0 - 2.0*I
```

smichr commented:
@jksuom , do you have any feedback about this?

jksuom commented:
I'm inclined to agree that some automatic simplifications like this would be acceptable and even useful. In general, there seems to be no reason for using an expression like `sqrt(3 + 4*I)` to denote a plain complex number with integer or rational components like `2 + I`. If that is really wanted, it could be constructed by setting `evaluate=False`.

This only affects a rather limited class of expressions. It seems to work also in some other cases like
```
>>> sqrt(1/(3 + 4*I))
2/5 - I/5
```
but not with fractions of type `(a + b*I)/(c + d*I)`:
```
>>> sqrt((3 + 4*I)/(3 - 4*I))
sqrt(-1/(3 - 4*I))*(1 - 2*I)
```
It looks like `1/(3 + 4*I)` is internally transformed to `(3 - 4*I)/25` at some point, but the corresponding transformation is not done on the fraction. It would be interesting to see if that could be arranged. Then this simplification would probably also work on fractions. But that need not be the subject of this PR.
Hence I am +1 for this.

asmeurer commented on sympy/core/add.py:
```diff
@@ -345,6 +345,25 @@ def as_coeff_Add(self, rational=False):
+            from sympy.core.evalf import pure_complex
+            from sympy.functions.elementary.complexes import sign
+            from sympy.functions.elementary.miscellaneous import sqrt
+            n, d = self.as_numer_denom()
```
What is the point of splitting out `d` here? Why not only operate on `pure_complex(self)`?

smichr replied:
I think I was trying to fail early. I just rewrote this. Let's see how the tests go.

asmeurer commented on sympy/core/tests/test_eval_power.py:
```diff
@@ -347,5 +347,25 @@ def test_issue_8582():
+    assert (x**(5*(n + 1))).is_positive is True
+
+
+def test_better_sqrt():
```
Should add some tests for the cases where it doesn't simplify.

smichr commented:
> It would be interesting to see if that could be arranged

I have modified the `Pow._eval_power` routine to look for power `1/(a+I*b)` and the `Mul._eval_power` routine to look for `n/(a+I*b)`. Tests were added and all added lines are covered.

@asmeurer , I now only process `self` instead of the numerator and denominator.

I have tried to avoid the error of treating sqrt(a/b) as sqrt(a)/sqrt(b) except when `a` or `b` is positive.

smichr commented:
ping @jksuom 
ping @asmeurer

jksuom commented on sympy/core/tests/test_eval_power.py:
```diff
@@ -347,5 +347,41 @@ def test_issue_8582():
+    assert sqrt((3 - 4*I)/25) == (2 - I)/5
+    assert sqrt((3 - 4*I)/26) == (2 - I)/sqrt(26)
+    # mul
+    assert sqrt((3 + 4*I)/(3 - 4*I)) == (1 - 2*I)/(1 + 2*I)  # not (2 + I)/(2 - I)
```
`(1 - 2*I)/(1 + 2*I)` has negative real part.

smichr replied:
:-( Hmmm. I'll have to see what happened with that.

smichr replied:
This is a consequence of how master (incorrectly) splits up the arguments:

```
>>> sqrt((3 + 4*I)/(3 - 4*I))
sqrt(-1/(3 - 4*I))*sqrt(-3 - 4*I)
```

smichr replied:
That is now fixed.

jksuom commented:
I wonder if there is any reason to leave the denominators in complex form. We could always transform `1/(a + b*I)` into `(a - b*I)/c` where `c = a**2 + b**2`. Then `sqrt((3 + 4*I)/(3 - 4*I))` would become

    (2 + I)/(2 - I) = (2 + I)**2/5 = 3/5 + 4/5*I

smichr commented:
> I wonder if there is any reason to leave the denominators in complex form

One reason might be to keep numbers smaller. e.g.  with powers more negative than -1, the numbers can become large and esthetically I find that less appealing:

```
(3 + 4*I)/(3 - 4*I)**3 == (-117 + 44*I)/15625*(3+4*I)
```

But let me know what you think. 

What I've got only expands when the power is -1 but that can be easily changed (by reverting the last commit). That will also simplify the `_eval` code, too. I think what I have is too special cased right now. If the simple `a + I*b` in the denominator automatically comes to the numerator then the only routine that needs changing is the `Add._eval_power`: for the case of exponent of -1 and the case when the exponent has a denominator of 2.

smichr committed:
keep unevaluated expressions unevaluated after normal

smichr committed:
drop `== True` use with Eq; use numer and denom

smichr committed:
modify _eval_power methods

o The holonomic test was modified because the string was failing
at the insignificant digits of the imaginary part (which is
just supposed to be small).

o The apart test was modified to pass

in master apart(f) gave I/((2 + I)*(x + I)) - 2*(-2 + I)/(5*(x - 2))
and now this becomes (in each step of the test):

>>> apart(f)
-I*(-2 + I)/(5*(x + I)) - 2*(-2 + I)/(5*(x - 2))
>>> together(_)
(-2/5 + I/5)*(-2*x - I*(x - 2) - 2*I)/((x - 2)*(x + I))
>>> factor(_)
-x*(-2 + I)*(2 + I)/(5*(x - 2)*(x + I))

o The rewriting of sqrt(1/i) in the test,

    i = symbols('i', imaginary=True)
    # this expansion of sqrt(1/i) happens at line 297 of power.py
    assert sqrt(3/i) == Mul(sqrt(3), sqrt(-i)/abs(i), evaluate=False)

happens in line 296 of power.py. If this is not desired, the condition
"and b.is_number" could be added to the end of line 295.

            # Please enter the commit message for your changes. Lines starting

smichr committed:
remove unused modifications to Mul._eval_power

smichr committed:
correct tests

smichr committed:
fully expand re and im parts in as_real_imag

smichr committed:
is_real_imaginary handles non-imaginary factors

smichr committed:
12775: simplify sqrt(2*r**2*I) as (r + I*r)

smichr commented:
> Trouble finding roots #12775

ping @jksuom , that is now added, too.

Other things related to real and imaginary calculation broke when bringing `a+I*b` out of the denominator. I think I have those fixed. If the tests pass this is ready for a final review.

jksuom commented on sympy/core/add.py:
```diff
@@ -350,6 +350,28 @@ def as_coeff_Add(self, rational=False):
+                    if D.is_Rational:
+                        # (r, i, D) is a Pythagorean triple
+                        root = sqrt(factor_terms((D - r)/2))
+                        return root*expand_multinomial((
```
Should we have also `root**e.p`?

smichr replied:
I'll have to look later today. Maybe that is the source of troubles I am having with apart and the prudnikov test ([see here](https://travis-ci.org/sympy/sympy/jobs/245227050) for example).

smichr committed:
fix tests

jksuom commented:
It seems that the failure is induced by the improved `.as_real_imag`:
```
>>> (-(3*2**pi)**(1/pi)).as_real_imag()
(-(3*2**pi)**(1/pi), 0)    # current master

>>> (-(3*2**pi)**(1/pi)).as_real_imag()
(-2*3**(1/pi), 0)    # this PR
```
This makes `d == 0` in `test_issue_10304` so that `d.is_comparable` is `True`.

smichr commented:
> the failure is induced by the improved

That failure I worked around, but if you check out the link you will see a failure in hyperexpand. THAT's the sore spot right now.

smichr committed:
raise root to e.p, too

smichr committed:
handle negative and non-Integer im parts

smichr committed:
split a into r, i once in as_real_imag

jksuom commented:
It seems that the failure in hyperexpand is the same as elsewhere, only it is invisible since bottom_up will catch the AttributeError.

smichr committed:
add note about exact in fraction docstring

smichr committed:
massage numerator of together's return value

Without this expansion and factoring, the return value of
together(apart(f)) with f = x/((x - 2)*(x + I)) would be

(-2/5 + I/5)*(-2*x - I*(x - 2) - 2*I)/((x - 2)*(x + I))

Anecdotally, there are also several expressions generated
by the test suite which come
back simplified to 0 that would otherwise have come back
as factored expressions.

smichr committed:
use expand_mul in carryout_plan

I am not sure if the division of an unexpanded expression
in a Poly would be better handled at the Poly level or if
the calling program should make sure that the Poly can be
manipulated as desired. This is the error that occurs if
the expansion is not used:

________ sympy\simplify\tests\test_hyperexpand.py:test_prudnikov_misc _________
  File "\sympy\sympy\simplify\tests\test_hyperexpand.py", line 663, in test_prudnikov_misc
    assert can_do([1, (3 + I)/2, (3 - I)/2], [S(3)/2, 2])
  File "\sympy\sympy\simplify\tests\test_hyperexpand.py", line 45, in can_do
    r = hyperexpand(hyper(ap, bq, z))
  File "sympy\simplify\hyperexpand.py", line 2471, in hyperexpand
    return f.replace(hyper, do_replace).replace(meijerg, do_meijer)
  File "sympy\core\basic.py", line 1407, in replace
    rv = bottom_up(self, rec_replace, atoms=True)
  File "sympy\simplify\simplify.py", line 998, in bottom_up
    rv = F(rv)
  File "sympy\core\basic.py", line 1392, in rec_replace
    new = _value(expr, result)
  File "sympy\core\basic.py", line 1335, in <lambda>
    _value = lambda expr, result: value(*expr.args)
  File "sympy\simplify\hyperexpand.py", line 2460, in do_replace
    r = _hyperexpand(Hyper_Function(ap, bq), z, rewrite=rewrite)
  File "sympy\simplify\hyperexpand.py", line 2039, in _hyperexpand
    r = carryout_plan(formula, ops) + p
  File "sympy\simplify\hyperexpand.py", line 1967, in carryout_plan
    C = C.applyfunc(make_simp(z0))
  File "sympy\matrices\common.py", line 1548, in applyfunc
    return self._eval_applyfunc(f)
  File "sympy\matrices\common.py", line 1488, in _eval_applyfunc
    out = self._new(self.rows, self.cols, [f(x) for x in self])
  File "sympy\simplify\hyperexpand.py", line 465, in simp
    c, numer, denom = poly(numer, z).cancel(poly(denom, z))

    >>> the Poly being created is poly((1+I)*(I-1),x), poly(2*x, x)

  File "sympy\polys\polytools.py", line 3646, in cancel
    result = F.cancel(G, include=include)
  File "sympy\polys\polyclasses.py", line 680, in cancel
    cF, cG, F, G = dmp_cancel(F, G, lev, dom, include=False)
  File "sympy\polys\euclidtools.py", line 1874, in dmp_cancel
    _, p, q = dmp_inner_gcd(f, g, u, K)
  File "sympy\polys\euclidtools.py", line 1588, in dmp_inner_gcd
    return dup_inner_gcd(f, g, K)
  File "sympy\polys\euclidtools.py", line 1520, in dup_inner_gcd
    return dup_ff_prs_gcd(f, g, K)
  File "sympy\polys\euclidtools.py", line 1038, in dup_ff_prs_gcd
    cff = dup_quo(f, h, K)
  File "sympy\polys\densearith.py", line 1575, in dup_quo
    return dup_div(f, g, K)[0]
  File "sympy\polys\densearith.py", line 1531, in dup_div
    return dup_ff_div(f, g, K)
  File "sympy\polys\densearith.py", line 1454, in dup_ff_div
    raise PolynomialDivisionFailed(f, g, K)
PolynomialDivisionFailed: couldn't reduce degree in a polynomial division algori
thm when dividing [EX((-1 + I)*(1 + I))] by [EX(-(-1 + I)*(1 - I)*(1 + I)**2/4)]
. This can happen when it's not possible to detect zero in the coefficient domai
n. The domain of computation is EX. You may want to use a different simplificati
on algorithm. Note that in general it's not possible to guarantee to detect zero
 in this domain.

smichr committed:
wrap long line

smichr committed:
give together the option to expand

smichr committed:
Revert "give together the option to expand"

This reverts commit a8dd4b36615affa2c5ac97dda93617d181ed3651.

smichr committed:
Revert "massage numerator of together's return value"

This reverts commit 67e08ac6dd721c8b8a0edf2bce6353d280f8c5c1.

smichr commented:
> failure in hyperexpand

I tracked down the failure and now use `expand_mul` in the simplification which aims at simplifying a rational function. Basically, something like `cancel((1+I)*(1-I)/(2*x))` is being attempted and the polys created from the numerator and denominator don't cancel without an error. If `Poly` were used instead of `poly` this wouldn't be an issue. I left a note in the commit message titled "use expand_mul in carryout_plan".

I thought that modifying `together` would be a good idea (since simpler `together` expressions would result) but the current naive behavior is expected so I just left the extra expansion in the `partfrac` test. Otherwise an option `expand` would have to be given to `together`. I did that and reverted it as shown in the history.

jksuom commented:
This looks like the old problem with polys over the EX domain: https://github.com/sympy/sympy/issues/12345, https://github.com/sympy/sympy/issues/5428. We never decided to enable extension=True by default, but in this case I suppose it could be set [here](https://github.com/sympy/sympy/blob/master/sympy/simplify/hyperexpand.py#L465) and probably also elsewhere in hyperexpand. That would make the simplification of expressions containing I automatic.

smichr commented:
> That would make the simplification of expressions containing I automatic.

I'm not sure that I need simplification in general...the `expand_mul` is simplification enough in this case. (And I don't have the imagination to think of cases where this would not be enough.) Is this too much of a hack or does this look ok to you? I could also send the `expand=True` keyword as in `poly(numer, expand=True)` but I'm not sure if that's overkill for this need. Apparently the person writing this code thought `poly` instead of `Poly` was going to be sufficient so I don't want to bog things down with unnecessary steps.

smichr committed:
use full expand on numer

Note: sending expand=True does not cause expand to be
used on the expression being passed to `poly`; it is passed
along to the factor programs.

smichr commented:
> .the expand_mul is simplification enough

No, it's not. A few lines down another test failed. I'm trying the `expand=True` flag now.

smichr commented:
It was the old "ball of string" with this one. I think all the issues have been resolved. This is ready for a final review.

asmeurer commented:
extension=True could have some implications. It's not enabled by default for performance purposes.

jksuom commented:
I have tested this with a version of `construct_domain` that yields `QQ<I>` instead of `EX` when there are complex coefficients but `extension=False` (essentially doing [this](https://github.com/sympy/sympy/blob/master/sympy/polys/constructor.py#L22)). It seems to work with no need to use `expand_mul` in `hyperexpand`.

asmeurer commented:
Conditionally doing extension=True just for `I` (but not yet general algebraic numbers) is probably a safer change. Ideally it would always be True, but (unless this has changed) there are still performance issues if you do that with more than a couple of extensions.

smichr commented:
The way it is right now, a full expand is needed on the numerator to pass all the `pudnikov` tests. @jksuom , do you want to do a PR to this or should this be committed as is?

smichr committed:
remove expand_mul import

jksuom commented:
This looks good to me on the whole. I only feel a bit uneasy about the `expand` in hyperexpand since it seems unrelated to the code. It masks an issue that should be taken care of at lower level, for example in `construct_domain` the way I had tested. But it may be too early to merge my tentative PR before more extensive tests. I think this could be merged now as it is.

smichr commented:
> . But it may be too early to merge my tentative PR

OK. I'll merge as is. Thanks for your help.
